Bad error message on invalid archive format

**Is this a BUG REPORT or FEATURE REQUEST?**:
Bug report

**What happened?**:
If pass a invalid file to --archive. We see this acr-builder blow up like this:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x81384a9]

goroutine 1 [running]:
compress/gzip.(*Reader).Close(0x0, 0x19444440, 0x194000e0)å
	/usr/local/go/src/compress/gzip/gunzip.go:292 +0x19
github.com/Azure/acr-builder/pkg/commands.unTAR.func1(0x0)
	/go/src/github.com/Azure/acr-builder/pkg/commands/archive.go:83 +0x21
github.com/Azure/acr-builder/pkg/commands.unTAR(0xffe25ad2, 0x15, 0xf7ef8600, 0x19781280, 0x83eab80, 0x1958a748)
	/go/src/github.com/Azure/acr-builder/pkg/commands/archive.go:89 +0x3d8
github.com/Azure/acr-builder/pkg/commands.(*ArchiveSource).Obtain(0x195379a0, 0x83ee1e0, 0x1958abd8, 0x0, 0x0)
	/go/src/github.com/Azure/acr-builder/pkg/commands/archive.go:48 +0x197
github.com/Azure/acr-builder/pkg.(Source).Obtain-fm(0x83ee1e0, 0x1958abd8, 0x0, 0x0)
	/go/src/github.com/Azure/acr-builder/pkg/driver/build.go:227 +0x33
github.com/Azure/acr-builder/pkg/workflow.(*Workflow).ScheduleRun.func1(0x83ee1e0, 0x1958abd8, 0x19537a00, 0x0, 0x0)
	/go/src/github.com/Azure/acr-builder/pkg/workflow/workflow.go:70 +0x2b
github.com/Azure/acr-builder/pkg/workflow.(*Workflow).Run(0x19537a00, 0x83ee1e0, 0x1958abd8, 0x0, 0x0)
	/go/src/github.com/Azure/acr-builder/pkg/workflow/workflow.go:49 +0x8b
github.com/Azure/acr-builder/pkg/driver.(*Builder).Run(0x194e5f7c, 0xffe25abb, 0x8, 0x0, 0x0, 0x0, 0x0, 0xffe25b1c, 0x1b, 0xffe25b47, ...)
	/go/src/github.com/Azure/acr-builder/pkg/driver/build.go:63 +0x2fe
main.main()
	/go/src/github.com/Azure/acr-builder/main.go:70 +0xc5f

Build ID: eastus-9 failed because of an internal error after 21.62147022s

**What did you expect to happen?**:
A better error message
**How do you reproduce it (as minimally and precisely as possible)?**:
acr-builder --archive **https://github.com/sajayantony/dockerfiles

--- a/pkg/commands/archive.go
+++ b/pkg/commands/archive.go
@@ -17,7 +17,7 @@ import (
 	"github.com/Azure/acr-builder/pkg/constants"
 )
 
-var maxHeaderSize = 4
+const maxHeaderSize = 4
 
 // NOTE: only support .tar.gz for now
 var supportedArchiveHeaders = map[byte][]byte{
@@ -56,7 +56,11 @@ func (s *ArchiveSource) Obtain(runner build.Runner) error {
 	if err != nil && err != io.EOF {
 		return fmt.Errorf("Failed to peek context header: %s", err)
 	}
-	if !isSupportedArchive(peek) {
+	supported, err := isSupportedArchive(peek)
+	if err != nil {
+		return fmt.Errorf("Failed to read context header: %s", err)
+	}
+	if !supported {
 		return fmt.Errorf("Unexpected file format for %s", s.url)
 	}
 
@@ -157,7 +161,13 @@ func unTAR(baseDir string, r io.Reader) error {
 	}
 }
 
-func isSupportedArchive(source []byte) bool {
+func isSupportedArchive(source []byte) (bool, error) {
+	if len(source) < 1 {
+		return false, fmt.Errorf("Empty header")
+	}
 	header, found := supportedArchiveHeaders[source[0]]
-	return found && bytes.Equal(source[:len(header)], header)
+	if len(source) < len(header) {
+		return false, fmt.Errorf("Format of the file content cannot be determined. File header is corrupted")
+	}
+	return found && bytes.Equal(source[:len(header)], header), nil
 }

--- a/pkg/commands/archive.go
+++ b/pkg/commands/archive.go
@@ -17,16 +17,12 @@ import (
 	"github.com/Azure/acr-builder/pkg/constants"
 )
 
-var maxHeaderSize = 512
-var supportedArchiveHeaders = map[byte][]byte{
-	// Bzip2
-	0x42: {0x42, 0x5A, 0x68},
+var maxHeaderSize = 4
 
+// NOTE: only support .tar.gz for now
+var supportedArchiveHeaders = map[byte][]byte{
 	// Gzip
 	0x1F: {0x1F, 0x8B, 0x08},
-
-	// Xz
-	0xFD: {0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00},
 }
 
 // ArchiveSource defines source in the form of an archive file
@@ -161,16 +157,7 @@ func unTAR(baseDir string, r io.Reader) error {
 	}
 }
 
-func isSupportedArchive(header []byte) bool {
-	if detectCompression(header) {
-		return true
-	}
-	r := tar.NewReader(bytes.NewBuffer(header))
-	_, err := r.Next()
-	return err == nil
-}
-
-func detectCompression(source []byte) bool {
+func isSupportedArchive(source []byte) bool {
 	header, found := supportedArchiveHeaders[source[0]]
 	return found && bytes.Equal(source[:len(header)], header)
 }

--- a/pkg/commands/archive.go
+++ b/pkg/commands/archive.go
@@ -19,9 +19,14 @@ import (
 
 var maxHeaderSize = 512
 var supportedArchiveHeaders = map[byte][]byte{
-	0x42: {0x42, 0x5A, 0x68},                   // Bzip2
-	0x1F: {0x1F, 0x8B, 0x08},                   // Gzip
-	0xFD: {0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, //Xz
+	// Bzip2
+	0x42: {0x42, 0x5A, 0x68},
+
+	// Gzip
+	0x1F: {0x1F, 0x8B, 0x08},
+
+	// Xz
+	0xFD: {0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00},
 }
 
 // ArchiveSource defines source in the form of an archive file
@@ -53,7 +58,7 @@ func (s *ArchiveSource) Obtain(runner build.Runner) error {
 	buf := bufio.NewReader(response.Body)
 	peek, err := buf.Peek(maxHeaderSize)
 	if err != nil && err != io.EOF {
-		return fmt.Errorf("failed to peek context header: %s", err)
+		return fmt.Errorf("Failed to peek context header: %s", err)
 	}
 	if !isSupportedArchive(peek) {
 		return fmt.Errorf("Unexpected file format for %s", s.url)
@@ -167,5 +172,5 @@ func isSupportedArchive(header []byte) bool {
 
 func detectCompression(source []byte) bool {
 	header, found := supportedArchiveHeaders[source[0]]
-	return found && bytes.Equal(source[0:len(header)], header)
+	return found && bytes.Equal(source[:len(header)], header)
 }

--- a/pkg/commands/archive.go
+++ b/pkg/commands/archive.go
@@ -2,6 +2,8 @@ package commands
 
 import (
 	"archive/tar"
+	"bufio"
+	"bytes"
 	"compress/gzip"
 	"fmt"
 	"io"
@@ -14,6 +16,13 @@ import (
 	build "github.com/Azure/acr-builder/pkg"
 	"github.com/Azure/acr-builder/pkg/constants"
 )
+
+var maxHeaderSize = 512
+var supportedArchiveHeaders = map[byte][]byte{
+	0x42: {0x42, 0x5A, 0x68},                   // Bzip2
+	0x1F: {0x1F, 0x8B, 0x08},                   // Gzip
+	0xFD: {0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, //Xz
+}
 
 // ArchiveSource defines source in the form of an archive file
 // Currently we only support tar.gz
@@ -41,11 +50,20 @@ func (s *ArchiveSource) Obtain(runner build.Runner) error {
 		}
 	}()
 
+	buf := bufio.NewReader(response.Body)
+	peek, err := buf.Peek(maxHeaderSize)
+	if err != nil && err != io.EOF {
+		return fmt.Errorf("failed to peek context header: %s", err)
+	}
+	if !isSupportedArchive(peek) {
+		return fmt.Errorf("Unexpected file format for %s", s.url)
+	}
+
 	baseDir := s.targetDir
 	if baseDir == "" {
 		baseDir = "."
 	}
-	if err := unTAR(baseDir, response.Body); err != nil {
+	if err := unTAR(baseDir, buf); err != nil {
 		return fmt.Errorf("Failed to extract archive from %s, error: %s", s.url, err)
 	}
 
@@ -136,4 +154,18 @@ func unTAR(baseDir string, r io.Reader) error {
 			logrus.Debugf("Ignoring unexpected file %s, type: %d", header.Name, header.Typeflag)
 		}
 	}
+}
+
+func isSupportedArchive(header []byte) bool {
+	if detectCompression(header) {
+		return true
+	}
+	r := tar.NewReader(bytes.NewBuffer(header))
+	_, err := r.Next()
+	return err == nil
+}
+
+func detectCompression(source []byte) bool {
+	header, found := supportedArchiveHeaders[source[0]]
+	return found && bytes.Equal(source[0:len(header)], header)
 }

--- a/pkg/commands/archive_test.go
+++ b/pkg/commands/archive_test.go
@@ -2,6 +2,7 @@ package commands
 
 import (
 	"context"
+	"net/http"
 	"os"
 	"path/filepath"
 	"regexp"
@@ -16,6 +17,7 @@ import (
 
 type obtainTestCase struct {
 	url               string
+	handler           http.Handler
 	targetDir         string
 	getWdErr          *error
 	expectedChdir     test.ChdirExpectations
@@ -29,7 +31,11 @@ func TestObtainFromKnownLocation(t *testing.T) {
 	targetDir := filepath.Join(testCommon.Config.ProjectRoot, "tests", "workspace")
 	testArchiveSource(t,
 		obtainTestCase{
-			url:       testCommon.StaticFileHost,
+			url: testCommon.StaticFileHost,
+			handler: &testCommon.StaticArchiveHandler{
+				T:           t,
+				ArchiveRoot: filepath.Join(testCommon.Config.ProjectRoot, "tests", "resources", "hello-multistage"),
+			},
 			targetDir: targetDir,
 			expectedChdir: []test.ChdirExpectation{
 				{Path: targetDir},
@@ -43,10 +49,30 @@ func TestObtainFromKnownLocation(t *testing.T) {
 	)
 }
 
+func TestObtainIncorrectFileFormat(t *testing.T) {
+	targetDir := filepath.Join(testCommon.Config.ProjectRoot, "tests", "workspace")
+	testArchiveSource(t,
+		obtainTestCase{
+			url: testCommon.StaticFileHost,
+			handler: &testCommon.StaticContentHandler{
+				T:       t,
+				Content: []byte("hello world"),
+			},
+			targetDir: targetDir,
+			expectedChdir: []test.ChdirExpectation{
+				{Path: targetDir},
+				{Path: "home"},
+			},
+			getWdErr:          &testCommon.NilError,
+			expectedObtainErr: "^Unexpected file format for .+",
+		},
+	)
+}
+
 func testArchiveSource(t *testing.T, tc obtainTestCase) {
 	cleanup(tc.targetDir)
 	defer cleanup(tc.targetDir)
-	server := testCommon.StartStaticFileServer(t)
+	server := testCommon.StartStaticFileServer(t, tc.handler)
 	defer testCommon.ReportOnError(t, func() error { return server.Shutdown(context.TODO()) })
 	source := NewArchiveSource(tc.url, tc.targetDir)
 	runner := test.NewMockRunner()

--- a/tests/testCommon/file_server.go
+++ b/tests/testCommon/file_server.go
@@ -19,10 +19,10 @@ const StaticFileHost = "http://localhost" + staticFilePort
 
 // StartStaticFileServer starts a file server, if error returned
 // is not null the caller should close it
-func StartStaticFileServer(t *testing.T) *http.Server {
+func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 	server := &http.Server{
 		Addr:    staticFilePort,
-		Handler: &staticFileArchiveHandler{},
+		Handler: handler,
 	}
 	go func() {
 		err := server.ListenAndServe()
@@ -55,25 +55,26 @@ func WaitForServerReady(timeoutInSeconds uint) error {
 	return fmt.Errorf("Server was still not ready after %v seconds", timeoutInSeconds)
 }
 
-// staticFileArchiveHandler streams the hello-multistage project to http response as tar.gz
+// StaticArchiveHandler streams the hello-multistage project to http response as tar.gz
 // We hard code the archive file for now, because we only have 1 test scenario using it
-type staticFileArchiveHandler struct {
-	t *testing.T
+type StaticArchiveHandler struct {
+	T           *testing.T
+	ArchiveRoot string
 }
 
-func (h *staticFileArchiveHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	projectRoot := filepath.Join(Config.ProjectRoot, "tests", "resources", "hello-multistage")
-	err := h.streamArchive(projectRoot, w)
+// ServeHTTP serves content of a directory
+func (h *StaticArchiveHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	err := h.streamArchive(h.ArchiveRoot, w)
 	if err != nil {
 		panic(err.Error())
 	}
 }
 
-func (h *staticFileArchiveHandler) streamArchive(root string, w io.Writer) error {
+func (h *StaticArchiveHandler) streamArchive(root string, w io.Writer) error {
 	gw := gzip.NewWriter(w)
-	defer ReportOnError(h.t, gw.Close)
+	defer ReportOnError(h.T, gw.Close)
 	tw := tar.NewWriter(gw)
-	defer ReportOnError(h.t, tw.Close)
+	defer ReportOnError(h.T, tw.Close)
 
 	return filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
@@ -96,8 +97,21 @@ func (h *staticFileArchiveHandler) streamArchive(root string, w io.Writer) error
 		if err != nil {
 			return err
 		}
-		defer ReportOnError(h.t, file.Close)
+		defer ReportOnError(h.T, file.Close)
 		_, err = io.Copy(tw, file)
 		return err
 	})
+}
+
+// StaticContentHandler serves a fixed string
+type StaticContentHandler struct {
+	T       *testing.T
+	Content []byte
+}
+
+// ServeHTTP serves the static contents
+func (h *StaticContentHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if _, err := w.Write(h.Content); err != nil {
+		panic(err)
+	}
 }

--- a/tests/testCommon/file_server.go
+++ b/tests/testCommon/file_server.go
@@ -24,15 +24,25 @@ func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 		Addr:    staticFilePort,
 		Handler: handler,
 	}
+	startRetryN := 10
+	startRetryInterval := time.Second * 10
 	go func() {
-		err := server.ListenAndServe()
-		if err != nil && err != http.ErrServerClosed {
-			t.Errorf("Failed to start server: %s", err)
-			t.Fail()
+		retryCount := 0
+		// NOTE: Sometimes when a test ends and a server shutdown port still appears to be used
+		for retryCount < startRetryN {
+			err := server.ListenAndServe()
+			if err != nil && err != http.ErrServerClosed {
+				t.Errorf("Failed to start server: %s", err)
+				t.Fail()
+				break
+			}
+			time.Sleep(startRetryInterval)
+			retryCount++
 		}
 	}()
 
-	err := WaitForServerReady(10)
+	totalDuration := startRetryInterval * (time.Duration)(startRetryN)
+	err := waitForServerReady(totalDuration, time.Second)
 	if err != nil {
 		t.Fail()
 	}
@@ -40,10 +50,10 @@ func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 	return server
 }
 
-// WaitForServerReady waits for server ready until timeout
-func WaitForServerReady(timeoutInSeconds uint) error {
+// waitForServerReady waits for server ready until timeout
+func waitForServerReady(timeout time.Duration, interval time.Duration) error {
 
-	for i := uint(0); i < timeoutInSeconds; i++ {
+	for remaining := timeout; remaining > 0; remaining -= interval {
 		time.Sleep(1 * time.Second)
 
 		_, err := http.Head(StaticFileHost)
@@ -52,7 +62,7 @@ func WaitForServerReady(timeoutInSeconds uint) error {
 		}
 	}
 
-	return fmt.Errorf("Server was still not ready after %v seconds", timeoutInSeconds)
+	return fmt.Errorf("Server was still not ready after %v seconds", timeout/time.Second)
 }
 
 // StaticArchiveHandler streams the hello-multistage project to http response as tar.gz

--- a/tests/testCommon/file_server.go
+++ b/tests/testCommon/file_server.go
@@ -28,17 +28,20 @@ func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 	startRetryInterval := time.Second * 10
 	go func() {
 		retryCount := 0
+		started := false
 		// NOTE: Sometimes when a test ends and a server shutdown port still appears to be used
-		for retryCount < startRetryN {
+		for !started && retryCount <= startRetryN {
 			err := server.ListenAndServe()
 			if err != nil && err != http.ErrServerClosed {
+				t.Errorf("Failed to start server: %s", err)
 				time.Sleep(startRetryInterval)
 				retryCount++
 			} else {
-				t.Errorf("Failed to start server: %s", err)
-				t.Fail()
-				break
+				started = true
 			}
+		}
+		if !started {
+			t.Fail()
 		}
 	}()
 

--- a/tests/testCommon/file_server.go
+++ b/tests/testCommon/file_server.go
@@ -32,12 +32,13 @@ func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 		for retryCount < startRetryN {
 			err := server.ListenAndServe()
 			if err != nil && err != http.ErrServerClosed {
+				time.Sleep(startRetryInterval)
+				retryCount++
+			} else {
 				t.Errorf("Failed to start server: %s", err)
 				t.Fail()
 				break
 			}
-			time.Sleep(startRetryInterval)
-			retryCount++
 		}
 	}()
 

--- a/tests/testCommon/file_server.go
+++ b/tests/testCommon/file_server.go
@@ -30,10 +30,10 @@ func StartStaticFileServer(t *testing.T, handler http.Handler) *http.Server {
 		retryCount := 0
 		started := false
 		// NOTE: Sometimes when a test ends and a server shutdown port still appears to be used
-		for !started && retryCount <= startRetryN {
+		for !started && retryCount < startRetryN {
 			err := server.ListenAndServe()
 			if err != nil && err != http.ErrServerClosed {
-				t.Errorf("Failed to start server: %s", err)
+				fmt.Fprintf(os.Stderr, "Failed to start server: %s\n", err)
 				time.Sleep(startRetryInterval)
 				retryCount++
 			} else {

